src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -502,42 +492,52 @@ internal static void CreateSymbolicLink(string path, string pathToTarget, bool i
                 Debug.Assert(success);
 
-                // We use PrintName(Offset|Length) instead of SubstituteName(Offset|Length) given that we don't want to return
-                // an NT path when the link wasn't created with such NT path.
-                // Unlike SubstituteName and GetFinalPathNameByHandle(), PrintName doesn't start with a prefix.
-                // Another nuance is that SubstituteName does not contain redundant path segments while PrintName does.
-                // PrintName can ONLY return a NT path if the link was created explicitly targeting a file/folder in such way.
-                //   e.g: mklink /D linkName \??\C:\path\to\target.
+                // We always use SubstituteName(Offset|Length) instead of PrintName(Offset|Length),
+                // the latter is just the display name of the reparse point and it can show something completely unrelated to the target.
 
                 if (rbSymlink.ReparseTag == Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK)
                 {
-                    int printNameOffset = sizeof(Interop.Kernel32.SymbolicLinkReparseBuffer) + rbSymlink.PrintNameOffset;
-                    int printNameLength = rbSymlink.PrintNameLength;
+                    int offset = sizeof(Interop.Kernel32.SymbolicLinkReparseBuffer) + rbSymlink.SubstituteNameOffset;
+                    int length = rbSymlink.SubstituteNameLength;
 
-                    Span<char> targetPath = MemoryMarshal.Cast<byte, char>(bufferSpan.Slice(printNameOffset, printNameLength));
-                    Debug.Assert((rbSymlink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) == 0 || !PathInternal.IsExtended(targetPath));
+                    Span<char> targetPath = MemoryMarshal.Cast<byte, char>(bufferSpan.Slice(offset, length));
 
-                    if (returnFullPath && (rbSymlink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) != 0)
+                    bool isRelative = (rbSymlink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) != 0;
+                    if (!isRelative)
                     {
-                        // Target path is relative and is for ResolveLinkTarget(), we need to append the link directory.
-                        return Path.Join(Path.GetDirectoryName(linkPath.AsSpan()), targetPath);
-                    }
+                        // Absolute target is in NT format and we need to clean it up before return it to the user.
+                        if (targetPath.StartsWith(PathInternal.UncNTPathPrefix.AsSpan()))
+                        {
+                            // We need to prepend the Win32 equivalent of UNC NT prefix.
+                            return Path.Join(PathInternal.UncPathPrefix.AsSpan(), targetPath.Slice(PathInternal.UncNTPathPrefix.Length)).ToString();
+                        }
 
-                    return targetPath.ToString();
+                        Debug.Assert(targetPath.StartsWith(PathInternal.NTPathPrefix.AsSpan()));
```

Comment:
If this assert fails, the assumption would be that Windows has a bug, correct? We should always expect the path at this point to _at least_ begin with `\??\`.

Reply:
> the assumption would be that Windows has a bug

No, I expect that this assertion warns us in case of a corner-case that we might not know that differs from our assumption of "all NT paths are  prefixed with \\??\\".

> We should always expect the path at this point to at least begin with \??\

That's what the assert is for.

Reply:
Ok yes, we are on the same page.

Reply:
It doesn't means that Windows has a bug, it means that we have it.